### ISSUE TITLE
[flutter_tools] Add support for Linux host ARM

### ISSUE THREAD

**Issue by kenthinson** [flutter/flutter#56992]
Flutter tool fails to build on Ubuntu Server ARM64 for Raspberry Pi. Just to be clear I am trying to run the SDK on the PI then install the app to a Android device.

## Steps to Reproduce

Brand new flash of Ubuntu 64bit for Raspberry Pi aka "ubuntu-20.04-preinstalled-server-arm64+raspi"

Following install instructions at: https://flutter.dev/docs/get-started/install/linux
`git clone https://github.com/flutter/flutter.git -b stable`
`export PATH="$PATH:`pwd`/flutter/bin"`
`flutter`
## Logs
```
Building flutter tool...
/home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/pub: line 49: /home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/dart: cannot execute binary file: Exec format error
/home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/pub: line 49: /home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/dart: Success
Error: Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (9 tries left)
/home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/pub: line 49: /home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/dart: cannot execute binary file: Exec format error
/home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/pub: line 49: /home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/dart: Success
Error: Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (8 tries left)
/home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/pub: line 49: /home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/dart: cannot execute binary file: Exec format error
/home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/pub: line 49: /home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/dart: Success
Error: Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (7 tries left)
/home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/pub: line 49: /home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/dart: cannot execute binary file: Exec format error
/home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/pub: line 49: /home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/dart: Success
Error: Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (6 tries left)
^C
```

```
ubuntu@ubuntu:~/development$ flutter doctor -v
Building flutter tool...
/home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/pub: line 49: /home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/dart: cannot execute binary file: Exec format error
/home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/pub: line 49: /home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/dart: Success
Error: Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (9 tries left)
/home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/pub: line 49: /home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/dart: cannot execute binary file: Exec format error
/home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/pub: line 49: /home/ubuntu/development/flutter/bin/cache/dart-sdk/bin/dart: Success
Error: Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (8 tries left)
```

**Comment by christopherfujino** [comment]
@kenthinson Hi, thanks for opening this. Unfortunately we don't support running the tool on arm architectures. I'm thinking the action point for this issue should be to make this explicit on the website system requirements https://flutter.dev/docs/get-started/install/linux

**Comment by kenthinson** [comment]
@christopherfujino Thanks for the information. Please feel free to close this after the docs are updated. I’ll leave it open as a reminder until then :)

**Comment by christopherfujino** [comment]
> @christopherfujino Thanks for the information. Please feel free to close this after the docs are updated. I’ll leave it open as a reminder until then :)

SGTM!

**Comment by christopherfujino** [comment]
@kenthinson are you still interested in using the flutter sdk on a raspberry pi? I'm going to be investigating the viability of this and would love to have another tester :)

**Comment by kenthinson** [comment]
@christopherfujino I would love to use it on Raspberry Pi! I have a Pi 4 I can test on. Please let me know if you get it working.

**Comment by christopherfujino** [comment]
> @christopherfujino I would love to use it on Raspberry Pi! I have a Pi 4 I can test on. Please let me know if you get it working. I’m happy to help test it.

I have a hacked together flutter tool that will run on arm, but I'm working on cross-compiling the engine artifacts (our engine build tooling doesn't support arm). I also filed bug to request an official arm build of the android sdk (the prebuilt debian arm `android-tools` looks like it won't cut it and I wasn't able to compile it from source on my rpi 3).

I should probably order an rpi 4, as having 1gb of RAM has been a bottleneck :) I'll keep this issue updated.

**Comment by kenthinson** [comment]
@christopherfujino Awesome! If I can be of any help let me know. My goal is to compile on the pi and install on an Android phone.

**Comment by HidenoriMatsubayashi** [comment]
@christopherfujino
cc: @christopherfujino 

I already have the aarch64 support patch. Also, it has been confirmed to work with Jetson Nano and Raspberry Pi 4. The dart patch of the Flutter SDK can be publicized immediately, but at the same time I think that infrastructure support is also necessary. For example, uploading the aarch64 rootfs or Flutter Engine aarch64 build image to the server.

I would like to cooperate with support, but what about?

![Screenshot from 2020-06-11 17-32-47](https://user-images.githubusercontent.com/62131389/86985056-2ca8c000-c1cb-11ea-9caa-489ac74d457c.png)

**Comment by NO-ob** [comment]
> > @christopherfujino I would love to use it on Raspberry Pi! I have a Pi 4 I can test on. Please let me know if you get it working. I’m happy to help test it.
> 
> I have a hacked together flutter tool that will run on arm, but I'm working on cross-compiling the engine artifacts (our engine build tooling doesn't support building from ARM). I also filed bug to request an official arm build of the android sdk (the prebuilt debian arm `android-tools` looks like it won't cut it and I wasn't able to compile it from source on my rpi 3).
> 
> I should probably order an rpi 4, as having 1gb of RAM has been a bottleneck :) I'll keep this issue updated.

Thought i'd mention it but at the off chance you have a nintendo switch you can get linux running on it quite easily. I think it's got better hardware than the pi4

**Comment by christopherfujino** [comment]
> Thought i'd mention it but at the off chance you have a nintendo switch you can get linux running on it quite easily. I think it's got better hardware than the pi4

That's good to know! Unfortunately, I'm not sure my wife would approve of me potentially bricking our animal crossing hub :P

**Comment by christopherfujino** [comment]
> Regarding this issue, I have a related suggestion that supports cross build for Linux ARM64/aarch64 targets.
> 
> I already have the patch of supporting aarch64 cross build on Linux x86_64. Also, it has been confirmed to work with Jetson Nano and Raspberry Pi 4. I can send you a PR (this patch) immediately, but at the same time I think that infrastructure support is also necessary. For example, uploading the aarch64 rootfs or Flutter Engine aarch64 build image to the server.
> 
> I would like to cooperate with support, but what about? Should I suggest another issue?

@HidenoriMatsubayashi I would to see your PR! You're right it would need infra updates, but I can help you with those parts.

**Comment by HidenoriMatsubayashi** [comment]
@christopherfujino 

Thank you. I sent https://github.com/flutter/flutter/pull/61221 once, so please confirm and let me discuss it.

**Comment by subjectdenied** [comment]
i'm a bit confused, but does this mean, with this pr it is possible to run flutter desktop linux apps on arm64, like pinephone for an example?

**Comment by HidenoriMatsubayashi** [comment]
@subjectdenied 
Yes, you are right. If you cross-build the Flutter app for ARM 64 on Host x86_64 CPU, build results will basically work on Host ARM64 such as Raspberry Pi4.

**Comment by christopherfujino** [comment]
@HidenoriMatsubayashi ahh, I see! So that is technically addressing https://github.com/flutter/flutter/issues/60678, linux arm as a target deployment platform.

**Comment by subjectdenied** [comment]
> @subjectdenied
> Yes, you are right. If you cross-build a Flutter app for ARM 64 on Host x86_64 CPU, build results will basically work on Host ARM64 such as Raspberry Pi4.

can you give me a hint about how you cross-compile, i have the aarch64 toolchain installed on fedora rawhide, however when using your command i get an error about wrong "Relocations in generic ELF (EM: 62)" and "file in wrong format". i also installed an aarrch64 sysroot

**Comment by anaisbetts** [comment]
I'd like to add my +1 to this issue, supporting this will not only unblock development on Raspberry Pi devices, but it would also make development on Windows ARM64 machines viable since I use WSL2 to run ARM64 Linux. Since web is a target that is supported regardless of what CPU you're running, getting the tools to run would enable working on UI stuff

**Comment by HidenoriMatsubayashi** [comment]
I've been working on supporting ARM64 Host.

**Comment by mukhtharcm** [comment]
@HidenoriMatsubayashi are you trying to enable building flutter apps **from** arm64 devices?

**Comment by HidenoriMatsubayashi** [comment]
@mukhtharcm 
Sorry for late reply. Yes, that's right. The main target was to run the Flutter app on an Arm64 Host, but now it is.

**Comment by hhk7734** [comment]
Thanks @HidenoriMatsubayashi .
Your PR is really helpful. :)

I tested on ODROID-C4(aarch64).
https://wiki.loliot.net/docs/lang/flutter/engine/flutter-app-for-linux-arm64

**Comment by pratham3213** [comment]
@christopherfujino  I am also having a rasberry pi 4 (8GB). I wanna be a tester too.Is it out for testing or it is under construction..?

**Comment by christopherfujino** [comment]
> @christopherfujino I am also having a rasberry pi 4. I wanna be a tester too.Is it out for testing or it is under construction..?

https://github.com/flutter/flutter/pull/61221/files should land on master...very soon :)

**Comment by christopherfujino** [comment]
Android target dev is blocked on Android Studio being available on Arm64 https://github.com/flutter/flutter/issues/75864

**Comment by christopherfujino** [comment]
Web target development is blocked on building dart2js in among the Flutter Engine artifacts https://github.com/flutter/flutter/issues/74928

**Comment by Leleocastro** [comment]
Hi guys, i'm not trying to develop in Raspberry, but in my Galaxy S20 phone. I could install Ubuntu 20.04, vscode, flutter, no issues in flutter doctor, but I couldnt debug the code. Like images:
![Screenshot_20210317-235459_VNC Viewer](https://user-images.githubusercontent.com/24722339/111566715-7004f680-877c-11eb-88a4-4a9f13656b3e.jpg)


![Screenshot_20210317-235025_VNC Viewer](https://user-images.githubusercontent.com/24722339/111566691-62e80780-877c-11eb-83a1-f8d907da755b.jpg)

**Comment by Leleocastro** [comment]
Some ideas?

**Comment by HidenoriMatsubayashi** [comment]
`dev_compiler` (dart2js) has not been included in the Flutter SDK yet. See also: https://github.com/flutter/flutter/issues/74928

**Comment by mukhtharcm** [comment]
> Hi guys, i'm not trying to develop in Raspberry, but in my Galaxy S20 phone. I could install Ubuntu 20.04, vscode, flutter, no issues in flutter doctor, but I couldnt debug the code. Like images:
> ![Screenshot_20210317-235459_VNC Viewer](https://user-images.githubusercontent.com/24722339/111566715-7004f680-877c-11eb-88a4-4a9f13656b3e.jpg)
> 
> ![Screenshot_20210317-235025_VNC Viewer](https://user-images.githubusercontent.com/24722339/111566691-62e80780-877c-11eb-83a1-f8d907da755b.jpg)

How are you Installing ubuntu on phone?
Termux?

**Comment by Leleocastro** [comment]
> How are you Installing ubuntu on phone?
> Termux?

Yeah, Termux and Andronix. The easiest way I found.
Download of Andronix: https://play.google.com/store/apps/details?id=studio.com.techriz.andronix

I connected to the GUI with VNC Viewer, starting vncserver on termux.
I use Samsung Dex to improve the experience. Now waiting for dart2js. (:

**Comment by christopherfujino** [comment]
Closing this as Linux Host ARM64 is supported for linux ARM64 and web targets on the latest master channel.
